### PR TITLE
Refactor and reorganize classes, add example and basic tests

### PR DESCRIPTION
**Pull request by luanpotter**
This is a very big PR, but I found no way to split this as everything depends on each other.
Upon using this package while working on [flame_noise](https://github.com/flame-engine/flame/pull/2393), I noticed several things that could be improved.

The main thing is that the *Noise classes did not share any interface so it was impossible to hold a reference to a generic Noise without using the noise2/noise3 helpers (which generate everything at once).

So this is what the PR does:

* create interfaces Noise2, Noise3, Noise2Int, Noise3Int (and less used Noise4 and Noise4Int). Most Noises are Noise2And3 (plus something else)
* separation of concerns: separate the Fractal versions of each noise into a separate calss. This makes it so the Noise classes maps 1:1 with NoiseType and make the application of the interface possible (for example, before Simplex could generate two types of noises so wouldn't fit on the interface; now it is broken into SimplexNoise and SimplexFractalNoise). also as part of separation of concerns, code related to gradiant perturbation is separated from CellularNoise as it was completely independent
* slim down each class to only contain the fields that it actually needs - some classes had unused fields
* extract a `buildNoise` method from `noise2`, which saves on loads of duplication and allow it to be used for other purposes as well
* put all code under `src` and expose everything under a single `fast_noise.dart` (before there was impl on the fast_noise.dart). I don't think it is big enough to have multiple libraries but happy to refactor that if desired
* replace example file with a proper Flutter app that easily allows for testing all types of 2d noise and combination of parameters (we can add 3d, etc as followups)
* add simple tests for each noise type
* adds flame_lint (a more strict linter) and fixes all violations
* re-order parameters and variables so they are more consistent across classes
* remove some code duplication (I believe there is still more to be improved on this regard)

### CODE REVIEW

File: pubspec.yaml [hunk 1]
```diff
@@ -2,7 +2,6 @@ name: fast_noise
 description: A port of the fast noise algorithms, featuring cellular_noise, cubic_noise, perlin_noise, simplex_noise, value_noise and white_noise
 version: 1.0.1
 homepage: https://github.com/frankpepermans/fast_noise
-author: Frank <frank@igindo.com>
```
Comment (luanpotter) [inline pubspec.yaml:5]: just to be clear, I removed this field because it is no longer supported in the pubspec format
https://dart.dev/tools/pub/pubspec

**Review by frankpepermans — APPROVED** [review]
Hey :)

Many thanks for this PR,
I'm not really supporting this package any more, so I'm open to transferring ownership.

Approved the PR (honestly been so long ago, don't even remember much of this package!)

**Comment by spydon** [comment]
> Hey :)
> 
> Many thanks for this PR, I'm not really supporting this package any more, so I'm open to transferring ownership.

Let's transfer it to Blue Fire then (Luan is also part of Blue Fire if that wasn't clear 😅).
I think what needs to be done then is:
1. Add one of us temporarily as admin to your [pub.dev publisher](https://pub.dev/publishers/igindo.com/packages) so that we can transfer it [to us](https://pub.dev/publishers/blue-fire.xyz/packages).
3. Add one of us as admins to this repository so that we can transfer it to the [Blue Fire GitHub organization](https://github.com/bluefireteam)

> Approved the PR (honestly been so long ago, don't even remember much of this package!)

I guess you could merge this PR before we transfer the package if you want to? :)

**Comment by luanpotter** [comment]
We (me and the rest of Blue Fire) would be delighted and honoured to help maintain it, @frankpepermans !

**Comment by frankpepermans** [comment]
Merge done!
Sorry on vacation atm, will return to this asap!

**Comment by spydon** [comment]
> Merge done!
> Sorry on vacation atm, will return to this asap!

Enjoy your vacation! :)

**Comment by frankpepermans** [comment]
@spydon I've added you to this repo and added your email to the pub organization (igindo.com)

**Comment by spydon** [comment]
> @spydon I've added you to this repo and added your email to the pub organization (igindo.com)

Great, thanks! I think you have to make me admin in the repo so that I can transfer the repository too.

**Pull request merged**